Providers/Profile.swift

```diff
@@ -576,28 +592,48 @@ public class BrowserProfile: Profile {
         }
 
         /**
-         * Locking is managed by withSyncInputs. Make sure you take and release these
+         * Locking is managed by syncSeveral. Make sure you take and release these
          * whenever you do anything Sync-ey.
          */
-        var syncLock = OSSpinLock() {
-            didSet {
-                let notification = syncLock == 0 ? NotificationProfileDidFinishSyncing : NotificationProfileDidStartSyncing
-                NSNotificationCenter.defaultCenter().postNotification(NSNotification(name: notification, object: nil))
-            }
-        }
+        private let syncLock = NSRecursiveLock()
 
-        // According to the OSAtomic header documentation, the convention for an unlocked lock is a zero value
-        // and a locked lock is a non-zero value
         var isSyncing: Bool {
-            return syncLock != 0
+            syncLock.lock()
+            defer { syncLock.unlock() }
+            return !(syncTerminal?.isFilled ?? true)
         }
 
+        private let syncQueue = dispatch_queue_create("com.mozilla.firefox.sync", DISPATCH_QUEUE_SERIAL)
+
         private func beginSyncing() -> Bool {
-            return OSSpinLockTry(&syncLock)
+            let ready = !isSyncing
+            if ready {
+                notifySyncing(NotificationProfileDidStartSyncing)
+            }
+            return ready
+        }
+
+        private func endSyncingMaybe(statuses: [EngineStatus] = []) {
+            // Sync `statuses` may have come from currentSync, or one before it.
+            syncLock.lock()
+            defer { syncLock.unlock() }
+            // If currentSync (the last sync job called) hasn't been filled, then we're still syncing.
```

Comment:
Lemme read this back, see if I'm understanding correctly.

We start with no syncs ongoing: both `currentSync` and `syncTerminal` are `nil`.

When a sync begins, we assign it to `currentSync` and put an empty `Deferred` into `syncTerminal`.

When another sync wants to join the queue, we chain it off `currentSync`, and assign the result to `currentSync`.

When any sync task finishes — i.e., fills a current or former value of `currentSync` – it calls `endSyncingMaybe`. This is guarded by a critical section so it can't run for more than one task at a time. If we're the last sync — the current sync has filled — then we're done.

We never leak `currentSync`, only `syncTerminal`; any post-sync tasks hang off the latter, so we can seamlessly switch in intermediates.

Right? If so, can we get a block comment a little like that up by those definitions? This logic is a little complicated!

(I don't have any good ideas to make it simpler, other than to encapsulate this concept of a task queue with a final result…)